### PR TITLE
823 traject - indexing locations and holdings

### LINKED ISSUE #823: Update locations and holding library in traject for Alma
In voyager a location is in one subfield. In Alma the location is a combination of two subfields.

For Alma we will index the combination into a single location field.

It's `852|b` and `852|c`.

https://github.com/pulibrary/marc_liberation/blob/a691bf10e8f475172a586921d003f8f280329589/marc_to_solr/lib/traject_config.rb#L1111-L1115

holding library: 

https://github.com/pulibrary/marc_liberation/blob/a691bf10e8f475172a586921d003f8f280329589/marc_to_solr/lib/traject_config.rb#L1142

### PR DESCRIPTION
**Pull request by christinach**
closes #823

### CODE REVIEW

**Comment by christinach** [comment]
I'm looking for an oversize holding id in Alma. @mzelesky said I can find one by looking for an 852 | k of 'Oversize'.

**Comment by christinach** [comment]
https://github.com/pulibrary/marc_liberation/pull/1033#issuecomment-771861761 This is to satisfy the process_holdings spec.

File: spec/marc_to_solr/lib/princeton_marc_spec.rb [hunk 1]
```diff
@@ -618,56 +627,64 @@
...
 
-    it 'excludes $c for call_number_browse key' do
-      expect(@holding_block[@oversize_mfhd_id]['call_number_browse']).not_to include('Oversize')
-      expect(@holding_block[@oversize_mfhd_id]['call_number_browse']).to eq(@call_number)
+    it 'positions $k at the end for call_number_browse field' do
+      expect(@oversize_holding['call_number_browse']).to include('Oversize')
+      expect(@oversize_holding['call_number_browse']).to eq("M23.L5S6 1973q Oversize")
     end
 
     it 'includes first instance of 852 $t as copy_number' do
```
Comment (christinach) [inline spec/marc_to_solr/lib/princeton_marc_spec.rb:678]: Is this still valid? To include the first instance of 852 $t as copy_number? The fixture I'm testing doesn't have a $t in the first 852. The code in https://github.com/pulibrary/marc_liberation/blob/cda1e2ce5466224748e6a74eab5ae55553faa059/marc_to_solr/lib/princeton_marc.rb#L653 doesn't check something like that. Do I want to update it or keep it as it is?

File: spec/marc_to_solr/lib/princeton_marc_spec.rb [hunk 2]
```diff
@@ -618,56 +627,64 @@
...
-      expect(@holding_block[@oversize_mfhd_id]['copy_number']).to eq('2')
+      expect(@oversize_holding['copy_number']).to eq('2')
     end
 
     it 'only includes copy_number if there is an 852 $t' do
       expect(@holding_block[@other_mfhd_id]['copy_number']).to be_nil
     end
```
Comment (christinach) [inline spec/marc_to_solr/lib/princeton_marc_spec.rb:684]: ok, I'm guessing the answer for https://github.com/pulibrary/marc_liberation/pull/1033/files#r569797757 is here. But I still want to confirm with @mzelesky .

File: spec/marc_to_solr/lib/princeton_marc_spec.rb [hunk 3]
```diff
@@ -618,56 +627,64 @@
...
 
     it 'separates call_number subfields with whitespace' do
-      expect(@holding_block[@oversize_mfhd_id]['call_number']).to eq("Oversize #{@call_number}")
+      expect(@oversize_holding['call_number']).to eq("M23.L5S6 1973q Oversize")
     end
```
Comment (christinach) [inline spec/marc_to_solr/lib/princeton_marc_spec.rb:688]: I need to update the code it adds 852|c in the call_number. In this case it adds 'stacks M23.L5S6 1973q Oversize'

File: spec/marc_to_solr/lib/princeton_marc_spec.rb (hunk 1 above)
Comment (christinach) [reply spec/marc_to_solr/lib/princeton_marc_spec.rb:678]: I will update this. In the checkin @mzelesky reminded me that the copy_number is part of the items now.

**Review by hackartisan — CHANGES_REQUESTED** [review]
@tpendragon and I looked at this together and left a couple of comments. Otherwise, I'd like to try it out locally; just need to get set up indexing to a local orangelight.

**Comment by hackartisan** [comment]
@christinach this will need to be rebased on the alma branch, since Monday happened.

I tried it out and I see a bunch of holdings locations coming through. I also see a bunch that say "no holdings available for this record". I'm not sure what those mean. But this does seem like progress! Here's a screenshot from my local orangelight:

Screen Shot 2021-02-15 at 1 46 35 PM

**Comment by christinach** [comment]
@hackartisan thanks. Will do.

File: marc_to_solr/lib/princeton_marc.rb [hunk 4]
```diff
@@ -649,13 +659,13 @@ def process_holdings record # rubocop:disable Metrics/AbcSize, Metrics/Cyclomati
     end
     holding['call_number'] = holding['call_number'].join(' ') if holding['call_number']
     holding['call_number_browse'] = holding['call_number_browse'].join(' ') if holding['call_number_browse']
-    all_holdings[holding_id] = holding unless holding_id.nil? || invalid_location?(holding['location_code'])
+    all_holdings[holding_id] = holding unless holding_id.nil? || (!is_alma && invalid_location?(holding['location_code']))
```
[the remark this answers is not in the document]
Comment (christinach) [reply marc_to_solr/lib/princeton_marc.rb:662]: @hackartisan  @tpendragon  The `is_alma` is checking if it is a valid princeton record. If we are positive that the records we have in the production box are all valid princeton records, I can remove the `is_alma` check. I don't understand how this is related to 'so we do want to make sure the location is valid'.

File: marc_to_solr/lib/princeton_marc.rb [hunk 5]
```diff
@@ -623,14 +628,19 @@ def process_holdings record # rubocop:disable Metrics/AbcSize, Metrics/Cyclomati
...
-      elsif /[ckhij]/.match?(s_field.code)
+        holding['location_code'] ||= s_field.value if is_alma
+        # Append 852c to location code 852b if it's an Alma item
+        holding['location_code'] += "$#{field['c']}" if field['c'] && is_alma
+        holding['location'] ||= Traject::TranslationMap.new("locations", default: "__passthrough__")[holding['location_code']] if is_alma
+        holding['library'] ||= Traject::TranslationMap.new("location_display", default: "__passthrough__")[holding['location_code']] if is_alma
+      elsif /[khij]/.match?(s_field.code)
         holding['call_number'] ||= []
         holding['call_number'] << s_field.value
         unless s_field.code == 'c'
```
[the remark this answers is not in the document]
Comment (christinach) [reply marc_to_solr/lib/princeton_marc.rb:646]: yes, sorry about that. I updated the subfields for the call_number and removed c from the match, but overlooked the unless branch. I will remove it.

File: marc_to_solr/lib/princeton_marc.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (christinach) [reply marc_to_solr/lib/princeton_marc.rb:660]: oh sorry now I see the exact line! yes I will remove it from `all_holdings[holding_id] = holding unless holding_id.nil? || (!is_alma && invalid_location?(holding['location_code']))`

File: spec/marc_to_solr/lib/princeton_marc_spec.rb [hunk 6]
```diff
@@ -625,60 +634,61 @@
...
+
+    it 'includes only first location code' do
+      expect(@oversize_holding_block['location_code']).to eq("annex$stacks")
     end
 
-    it 'excludes $c for call_number_browse key' do
-      expect(@holding_block[@oversize_mfhd_id]['call_number_browse']).not_to include('Oversize')
-      expect(@holding_block[@oversize_mfhd_id]['call_number_browse']).to eq(@call_number)
+    # TODO: ALMA
+    xit 'excludes holdings with an invalid location code' do
```
[the remark this answers is not in the document]
Comment (christinach) [reply spec/marc_to_solr/lib/princeton_marc_spec.rb:661]: Ok, I will update the spec customizing the new fixture.

**Comment by christinach** [comment]
@hackartisan can you review again?

**Comment by christinach** [comment]
> @christinach this will need to be rebased on the alma branch, since Monday happened.
> 
> I tried it out and I see a bunch of holdings locations coming through. I also see a bunch that say "no holdings available for this record". I'm not sure what those mean. But this does seem like progress! Here's a screenshot from my local orangelight:
> 
> Screen Shot 2021-02-15 at 1 46 35 PM

The one that says no holdings available is an electronic record https://princeton.alma.exlibrisgroup.com/ng/page;u=%2Frep%2Faction%2FpageAction.search.mms_results.xml.do%3FxmlFileName%3Dsearch.mms_results.xml&pageViewMode%3DEdit&clear%3Dtrue&pageBean.securityHashToken%3D8857301989828534662
I exported it. It has an 856 : `<datafield tag="856" ind1="4" ind2="0"><subfield code="u">https://ebookcentral.proquest.com/lib/Princeton/detail.action?docID=6135628</subfield></datafield>`

It should display `ebookcentral.proquest.com`. From https://catalog-alma-qa.princeton.edu/catalog/99118919643506421/raw, 
"electronic_access_1display": "{\"https://ebookcentral.proquest.com/lib/Princeton/detail.action?docID=6135628\":[\"ebookcentral.proquest.com\"]}",

This is how it displays in the current catalog https://catalog.princeton.edu/catalog?utf8=%E2%9C%93&search_field=all_fields&q=Acceleration+hours+
Screen Shot 2021-02-17 at 12 46 28 PM

**Review by hackartisan — APPROVED** [review]

**Pull request merged**